azure/services/natgateways/natgateways.go

```diff
@@ -18,156 +18,112 @@ package natgateways
+	client := newClient(scope)
 	return &Service{
-		Scope:  scope,
-		client: newClient(scope),
+		Scope:      scope,
+		Reconciler: async.New(scope, client, client),
 	}
 }
 
-// Reconcile gets/creates/updates a nat gateway.
-// Only when the Nat Gateway 'Name' property is defined we create the Nat Gateway: it's opt-in.
+// Reconcile gets/creates/updates a NAT gateway.
+// Only when the NAT gateway 'Name' property is defined we create the NAT gateway: it's opt-in.
 func (s *Service) Reconcile(ctx context.Context) error {
 	ctx, log, done := tele.StartSpanWithLogger(ctx, "natgateways.Service.Reconcile")
 	defer done()
 
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
 	if !s.Scope.Vnet().IsManaged(s.Scope.ClusterName()) {
 		log.V(4).Info("Skipping nat gateways reconcile in custom vnet mode")
+
+		s.Scope.UpdatePutStatus(infrav1.NATGatewaysReadyCondition, serviceName, nil)
 		return nil
 	}
 
+	// We go through the list of NatGatewaySpecs to reconcile each one, independently of the resultingErr of the previous one.
+	// If multiple errors occur, we return the most pressing one
+	// order of precedence is: error creating -> creating in progress -> created (no error)
+	var resultingErr error
 	for _, natGatewaySpec := range s.Scope.NatGatewaySpecs() {
-		existingNatGateway, err := s.getExisting(ctx, natGatewaySpec)
-
-		switch {
-		case err != nil && !azure.ResourceNotFound(err):
-			return errors.Wrapf(err, "failed to get nat gateway %s in %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
-		case err == nil:
-			// nat gateway already exists
-			log.V(4).Info("nat gateway already exists", "nat gateway", natGatewaySpec.Name)
-			natGatewaySpec.Subnet.NatGateway.ID = existingNatGateway.ID
-
-			if existingNatGateway.NatGatewayIP.Name == natGatewaySpec.NatGatewayIP.Name {
-				// Skip update for Nat Gateway as it exists with expected values
-				log.V(4).Info("Nat Gateway exists with expected values, skipping update", "nat gateway", natGatewaySpec.Name)
-				natGatewaySpec.Subnet.NatGateway = *existingNatGateway
-				s.Scope.SetSubnet(natGatewaySpec.Subnet)
-				continue
-			} else {
-				log.V(2).Info("updating NAT gateway IP name to match the spec", "old name", existingNatGateway.NatGatewayIP.Name, "desired name", natGatewaySpec.NatGatewayIP.Name)
+		result, err := s.CreateResource(ctx, natGatewaySpec, serviceName)
+		if err != nil {
+			if !azure.IsOperationNotDoneError(err) || resultingErr == nil {
+				resultingErr = err
 			}
-		default:
-			// nat gateway doesn't exist but its name was specified in the subnet, let's create it
-			log.V(2).Info("nat gateway doesn't exist yet, creating it", "nat gateway", natGatewaySpec.Name)
 		}
+		if err == nil {
+			natGateway, ok := result.(network.NatGateway)
+			if !ok {
+				// Return out of loop since this would be an unexpcted fatal error
+				resultingErr = errors.Errorf("created resource %T is not a network.NatGateway", result)
```

Comment:
```suggestion
				return errors.Errorf("created resource %T is not a network.NatGateway", result)
```

Reply:
I used the `break` statement to avoid calling `UpdatePutStatus` again similar to the changes in #1861, do you think it's worth changing back?

Reply:
got it